src/main/java/dev/jbang/AliasUtil.java

```diff
@@ -260,49 +423,107 @@ public static Aliases getCatalogAliasesByRef(String catalogRef, boolean updateCa
+		if (dir.getParent() != null) {
+			allAliasesFromLocalCatalogs(dir.getParent(), result);
+		}
+		Path catalog = dir.resolve(JBANG_DOT_DIR).resolve(JBANG_CATALOG_JSON);
+		if (Files.isRegularFile(catalog) && Files.isReadable(catalog)) {
+			result.aliases.putAll(getAliasesFromCatalogFile(catalog, false).aliases);
 		}
-		if (aliases.baseRef != null && !isAbsoluteRef(alias.scriptRef)) {
-			String ref = aliases.baseRef;
-			if (!ref.endsWith("/")) {
-				ref += "/";
+		catalog = dir.resolve(JBANG_CATALOG_JSON);
+		if (Files.isRegularFile(catalog) && Files.isReadable(catalog)) {
+			result.aliases.putAll(getAliasesFromCatalogFile(catalog, false).aliases);
+		}
+	}
+
+	private static Path findNearestLocalCatalog(Path dir) {
+		if (dir == null) {
+			dir = getCwd();
+		}
+		while (dir != null) {
+			Path catalog = dir.resolve(JBANG_CATALOG_JSON);
+			if (Files.isRegularFile(catalog) && Files.isReadable(catalog)) {
+				return catalog;
 			}
-			if (alias.scriptRef.startsWith("./")) {
-				ref += alias.scriptRef.substring(2);
-			} else {
-				ref += alias.scriptRef;
+			catalog = dir.resolve(JBANG_DOT_DIR).resolve(JBANG_CATALOG_JSON);
+			if (Files.isRegularFile(catalog) && Files.isReadable(catalog)) {
+				return catalog;
 			}
-			alias = new Alias(ref, alias.description, alias.arguments, alias.properties);
+			dir = dir.getParent();
 		}
-		// TODO if we have to combine the baseUrl with the scriptRef
-		// we need to make a copy of the Alias with the full URL
-		return alias;
+		return null;
+	}
+
+	public static Path findNearestLocalCatalogWithAlias(Path dir, String aliasName) {
+		if (dir == null) {
+			dir = getCwd();
+		}
+		while (dir != null) {
+			Path catalog = dir.resolve(JBANG_CATALOG_JSON);
+			if (Files.isRegularFile(catalog) && Files.isReadable(catalog)) {
+				Aliases aliases = getAliasesFromCatalogFile(catalog, false);
+				if (aliases.aliases.containsKey(aliasName)) {
+					return catalog;
+				}
+			}
+			catalog = dir.resolve(JBANG_DOT_DIR).resolve(JBANG_CATALOG_JSON);
+			if (Files.isRegularFile(catalog) && Files.isReadable(catalog)) {
+				Aliases aliases = getAliasesFromCatalogFile(catalog, false);
+				if (aliases.aliases.containsKey(aliasName)) {
+					return catalog;
+				}
+			}
+			dir = dir.getParent();
+		}
+		return null;
 	}
 
-	static Aliases readAliasesFromCatalog(Path catalogPath) {
+	public static Aliases getAliasesFromCatalogFile(Path catalogPath, boolean updateCache) {
 		Aliases aliases;
-		aliases = new Aliases();
+		if (updateCache || !catalogCache.containsKey(catalogPath)) {
+			aliases = readAliasesFromCatalogFile(catalogPath);
+			aliases.catalogFile = catalogPath;
+			catalogCache.put(catalogPath, aliases);
+		} else {
+			aliases = catalogCache.get(catalogPath);
+		}
+		return aliases;
+	}
+
+	static private Path scriptParent(Path script) {
+		Path parent = script.getParent();
+		if (parent.getFileName().toString().equals(JBANG_DOT_DIR)) {
```

Comment:
I still think we should have the actual folder of the jbang-catalog being what paths are resolved against. It just feels weird other wise.

can you point to a test/usecase that breaks/doesn't work for this ?